### PR TITLE
PR-A1b3 — D-1 layout walk U20 layout_dirty_root + disjoint-borrow recursion

### LINKED ISSUE #2: [CRITICAL] Rebuild pipeline is a complete no-op stub
## Problem

The `rebuild_dirty()` function in `build_pipeline.rs` lines 427-437 is a complete stub that does nothing but log. This is **production code masquerading as working implementation**.

```rust
// TODO(2025-11): Full View-based rebuilding
// For now, just log that we processed the element
// THE REBUILD FUNCTION DOESN'T ACTUALLY REBUILD ANYTHING!
```

## Impact

🔴 **CRITICAL - Show Stopping Bug**

- `setState()` calls do nothing
- State changes don't trigger UI updates  
- UI is completely static
- Completely breaks stateful widgets in production

## Location

- **File**: `crates/flui_core/src/pipeline/build_pipeline.rs`
- **Lines**: 427-437
- **Function**: `rebuild_dirty()`

## Recommended Fix

Either:
1. **Implement the full rebuild logic** per the TODO comment
2. **Add panic in debug mode** to prevent usage:
   ```rust
   #[cfg(debug_assertions)]
   panic!("rebuild_dirty() stub called - View::rebuild() not yet implemented");
   ```
3. **Mark API as unstable** with compile-time warnings

## Severity Justification

This is marked CRITICAL because it silently breaks core functionality. Developers will call `setState()` expecting UI updates, but nothing happens. This is worse than a crash - it's silent incorrect behavior.

## Related

Part of Widget → View migration. See TODO(2025-11) for implementation plan.

### LINKED ISSUE #3: [CRITICAL] Use-after-free risk in layout_render_object
## Problem

The `layout_render_object` method in `element_tree.rs` holds a reference to an element, then performs potentially mutating operations while still using that reference. If the element is removed from the tree during layout, this reference could become dangling.

```rust
let element = self.get(element_id)?;  // Line 421 - borrowed here
let render_element = element.as_render()?; // Line 422

// ... many operations later ...

// Line 485: Still using render_element obtained from old borrow
let size = render_element.render_object_mut().layout(self, constraints);
```

## Impact

🔴 **CRITICAL - Memory Safety**

- Potential **undefined behavior** if tree structure changes during layout
- Layout can trigger callbacks that may mutate the tree
- Re-entrant layout is already detected (lines 458-478), suggesting this is a real concern
- Could lead to crashes or memory corruption

## Location

- **File**: `crates/flui_core/src/pipeline/element_tree.rs`
- **Lines**: 421-505
- **Method**: `layout_render_object()`

## Recommended Fix

Re-fetch the element reference after any potentially mutating operation:

```rust
pub fn layout_render_object(&self, element_id: ElementId, constraints: BoxConstraints) 
    -> Option<flui_types::Size> {
    // Check cache first (this part is OK)
    {
        let element = self.get(element_id)?;
        let render_element = element.as_render()?;
        let render_state = render_element.render_state();
        let state = render_state.read();
        if state.has_size() && !state.needs_layout() 
            && state.constraints() == Some(constraints) {
            return state.size();
        }
    } // Drop all references here
    
    // Re-fetch for layout
    let element = self.get(element_id)?;
    let render_element = element.as_render()?;
    // ... continue with layout
}
```

## Severity Justification

Memory safety violations in Rust are always CRITICAL as they can lead to undefined behavior. The re-entrant layout detection code suggests this scenario is realistic, not just theoretical.

### PR DESCRIPTION
**Pull request by vanyastaff**
# PR-A1b3 / U20 — `PipelineOwner::layout_dirty_root` disjoint-borrow walk

Third sub-PR in the PR-A1b series (D-1 layout pipeline wiring). Follows [PR #139](https://github.com/vanyastaff/flui/pull/139) (PR-A1a — OnceCell→Option + mark_needs_layout walk), [PR #140](https://github.com/vanyastaff/flui/pull/140) (PR-A1b1 — protocol-erased dispatch), [PR #141](https://github.com/vanyastaff/flui/pull/141) (PR-A1b2 — typed bridge + Proxy ctx), and [PR #143](https://github.com/vanyastaff/flui/pull/143) (`perform_layout_raw` Result-returning signature).

## What this PR adds

`PipelineOwner::<Layout>::layout_dirty_root(id, constraints) -> RenderResult<Size>` — the production disjoint-borrow layout walk. Drives per-node `perform_layout_raw` against a Direct-storage `BoxLayoutCtx` populated with the parent's child IDs + a recursive `layout_subtree_raw` callback. Replaces the no-op recursion shape of `layout_node_with_children`.

Per plan §U20 + memo D1 ([`docs/plans/2026-05-23-001-feat-pipeline-wiring-d-block-plan.md`](docs/plans/2026-05-23-001-feat-pipeline-wiring-d-block-plan.md), [`docs/research/2026-05-23-d-block-architecture-decision-memo.md`](docs/research/2026-05-23-d-block-architecture-decision-memo.md)).

## Mechanism

Each `layout_subtree_raw` invocation:

1. Snapshots `child_ids` from the tree.
2. **Leaf path**: delegates to `RenderEntry::layout_leaf_only` (same U18/U19 leaf bridge code).
3. **Non-leaf path**: builds `Vec<ChildState<BoxParentData>>`, constructs `BoxLayoutCtx::with_layout_callback` with a recursive callback that re-enters `layout_subtree_raw` for each child slot, invokes `perform_layout_raw` wrapped in `std::panic::catch_unwind` (review fix — symmetric with leaf path).
4. On success **AND when no descendant errored**: updates `state.set_geometry` / `state.set_constraints`, bootstraps `IS_RELAYOUT_BOUNDARY`, clears `NEEDS_LAYOUT`. When a descendant errored mid-callback: state still recorded but `NEEDS_LAYOUT` stays SET for next-frame retry (review fix).

## Commits (atomic-per-unit shape)

| Commit | Summary |
|--------|---------|
| `4967a9c0` | U20a — add `RenderError::ChildIndexOutOfBounds` variant (reserved for future defensive checks) |
| `e9e697ff` | U20b — `PipelineOwner::layout_dirty_root` + `TreePtr` + `layout_subtree_raw` recursion helper |
| `e87dd3d7` | U20c — 6 integration tests (2-level, 3-level, leaf delegation, frame-2 idempotence, stale-id, RenderViewAdapter) |
| `026e3798` | Review fixes — `catch_unwind` on non-leaf; retain `NEEDS_LAYOUT` on descendant `Err`; distinguish `NodeNotFound` vs `ProtocolMismatch`; doc gating (`#[doc(hidden)]` + KNOWN MIRI BLOCKER note); 3 regression tests |

## Test scope

[`crates/flui-rendering/tests/u20_layout_dirty_root.rs`](crates/flui-rendering/tests/u20_layout_dirty_root.rs) — 9 tests:

- `u20_two_level_padding_with_colored_box_child` (plan happy path)
- `u20_three_level_padding_center_colored_box_grandchild_propagation` (plan edge case — recursive callback)
- `u20_stale_root_id_returns_node_not_found` (plan failure path)
- `u20_leaf_path_delegates_to_layout_leaf_only` (smoke)
- `u20_double_layout_does_not_panic_on_frame_two` (U14 OnceCell→Option regression smoke; AE8 surface)
- `u20_render_view_adapter_layout_smoke` (manual non-blanket `RenderObject<BoxProtocol>` impl smoke)
- `u20_non_leaf_perform_layout_panic_surfaces_as_poisoned` (review fix #1)
- `u20_descendant_err_preserves_parent_needs_layout` (review fix #2)
- `u20_sliver_node_surfaces_as_protocol_mismatch` (review fix #3)

## Known limitations — documented inline, blockers for U23 wiring

1. **KNOWN MIRI BLOCKER (latent UB)** — the recursive raw-pointer reborrow pattern in `layout_subtree_raw` holds an outer `&mut RenderEntry` borrow live across `perform_layout_raw`, during which the synchronous callback re-enters and synthesises a fresh `&mut RenderTree` from the same `*mut`. Stacked Borrows / Tree Borrows would flag this; `cargo +nightly miri test -p flui-rendering --test u20_layout_dirty_root` would fail. Current rustc/LLVM does not miscompile, but the pattern is latently unsound. **Architectural fix path**: pre-acquire all subtree `&mut RenderNode` borrows upfront via new `RenderTree::get_subtree_mut` primitive. **Blocker for U23's `run_layout` wiring** — this PR adds `#[doc(hidden)]` to keep the method un-promoted until the fix lands.

2. **No cycle guard / depth limit** — U21's job (`RenderError::LayoutCycle` variant already reserved). Until U21 lands, callers must ensure the tree is acyclic.

3. **ParentData scope** — pipeline-side `BoxLayoutCtx` is hardcoded to `BoxParentData`. Non-default ParentData types (`FlexParentData` on `RenderFlex`) trigger debug-assert in `BoxLayoutCtx::from_erased` in debug builds; silent wrong layout in release. Pipeline-driven flex is Core.1 follow-up.

4. **Descendant `Err` observability** — callback collapses descendant `RenderError` to `Size::ZERO`; tracing logs the failure; parent stays dirty for retry. Surfacing typed `Err` to the outer caller requires widening `LayoutChildCallback` to `Result` — Core.1 API-shape change.

## Code review (Tier 1)

15 findings across 5 finder angles (line-by-line / removed-behavior / cross-file / Rust-pitfall / wrapper-correctness). 3 confirmations via single-vote verifiers. Top 4 P0/P1/P2 findings addressed in commit `026e3798`; the remaining 11 either are documented inline on `layout_dirty_root` as known limitations (Miri UB, cycle guard, ParentData scope, `Err` observability) or are deferred to U23-wiring / Core.1 work as called out in the commit body.

## Verification

```
cargo build -p flui-rendering --lib                                           # clean
cargo clippy --workspace --all-targets -- -D warnings                          # clean
cargo test -p flui-rendering --lib                                             # 266 passed
cargo test -p flui-rendering --test u19_bridge                                 # 9 passed (no regression)
cargo test -p flui-rendering --test u20_layout_dirty_root                      # 9 passed
bash scripts/port-check.sh -v                                                  # all 7 triggers + FR-033 + FR-036 clean
```

## Post-Deploy Monitoring & Validation

No additional operational monitoring required — this is internal rendering pipeline work; `layout_dirty_root` is `#[doc(hidden)]` and not yet wired into any production caller (U23 will wire `run_layout`). Tests are the validation surface.

## Known Residuals

None accepted via residual gate — this PR ran Tier 1 review only (concentrated diff, no sensitive surface).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/flui-rendering/src/pipeline/owner.rs [hunk 1]
```diff
@@ -981,6 +990,447 @@ impl PipelineOwner<Layout> {
...
+        // doc above. The recursive call's borrows are on
+        // `child_id`'s slab slot (and its descendants), disjoint
+        // from the outer call's `id`-slot borrow.
+        match unsafe { layout_subtree_raw(tree_ptr, child_id, child_constraints) } {
+            Ok(size) => size,
+            Err(err) => {
+                // Set the flag so stage 6 preserves the parent's
+                // NEEDS_LAYOUT for next-frame retry.
+                descendant_error_for_cb.store(true, std::sync::atomic::Ordering::Relaxed);
+                tracing::error!(
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/flui-rendering/src/pipeline/owner.rs:1377]: Acknowledged as documented partial-mitigation. The descendant-error AtomicBool flag in 026e3798 ensures the parent stays NEEDS_LAYOUT-dirty for next-frame retry (so the outer Ok return is not a permanent silent success — subsequent walks re-run the failing subtree). Full propagation of typed Err to the outer caller requires widening LayoutChildCallback from `Fn(_) -> Size` to `Fn(_) -> Result<Size, _>` — Core.1 API-shape change documented inline on `layout_dirty_root` rustdoc. Not fixing further in this PR per scope decision; tracked as part of the U20.1 follow-up that gates U23 wiring.

File: crates/flui-rendering/src/pipeline/owner.rs [hunk 2]
```diff
@@ -981,6 +990,447 @@ impl PipelineOwner<Layout> {
...
+    // below. Per the helper-level safety doc invariant 3, the parent
+    // and child slots are distinct so the disjoint-borrow discipline
+    // holds across the recursion boundary (see the KNOWN MIRI BLOCKER
+    // note on `layout_dirty_root` for the latent UB caveat).
+    let cb_owned = move |child_id: RenderId, child_constraints: BoxConstraints| -> Size {
+        // SAFETY: see helper-level safety doc + the closure's own
+        // doc above. The recursive call's borrows are on
+        // `child_id`'s slab slot (and its descendants), disjoint
+        // from the outer call's `id`-slot borrow.
+        match unsafe { layout_subtree_raw(tree_ptr, child_id, child_constraints) } {
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/flui-rendering/src/pipeline/owner.rs:1371]: Acknowledged — this is the KNOWN MIRI BLOCKER documented prominently on `layout_dirty_root` rustdoc (see "Soundness" section). The architectural fix (pre-acquire all subtree `&mut RenderNode` upfront via new `RenderTree::get_subtree_mut` primitive, eliminating recursive raw-pointer reborrows entirely) is filed as the U20.1 follow-up and is an explicit blocker for U23s `run_layout` wiring. Method is `#[doc(hidden)]` until U20.1 lands. Not fixing in this PR per scope decision (Option A from the pre-push escalation).

File: crates/flui-rendering/src/pipeline/owner.rs [hunk 3]
```diff
@@ -981,6 +990,447 @@ impl PipelineOwner<Layout> {
...
+
+// SAFETY: the raw pointer is just an address; the per-call-site
+// `&mut RenderTree` synthesised inside [`layout_subtree_raw`] is the load-
+// bearing borrow, and that borrow is scoped to a single statement at each
+// use site. No data races are possible because the pipeline phase holds
+// `&mut self` for the entire layout walk and the callback fires
+// synchronously from inside `perform_layout_raw` on the same thread.
+unsafe impl Send for TreePtr {}
+// SAFETY: the same reasoning as `Send` — see above.
+unsafe impl Sync for TreePtr {}
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/flui-rendering/src/pipeline/owner.rs:1231]: Fixed in da4dc936: TreePtr now stores the owning ThreadId at construction (via TreePtr::new) and panics with a documented diagnostic if dereferenced from another thread (check_thread() called at the head of every layout_subtree_raw invocation). The guard is active in all builds (one ThreadId compare on the hot path). The structural fix (remove Send + Sync from BoxLayoutCtxErased — touches U19 design) is deferred to the same U20.1 follow-up as the recursive-reborrow fix.

File: crates/flui-rendering/src/pipeline/owner.rs [hunk 4]
```diff
@@ -981,6 +990,447 @@ impl PipelineOwner<Layout> {
...
+    // SAFETY: see helper-level safety doc invariant 2+3 AND the known
+    // soundness gap documented on `layout_dirty_root`.
+    //
+    // Review fix: split NodeNotFound vs ProtocolMismatch (same shape as
+    // the leaf-path acquisition above).
+    let parent_entry: &mut RenderEntry<BoxProtocol> = unsafe {
+        let node = match (*ptr).get_mut(id) {
+            Some(n) => n,
+            None => return Err(crate::error::RenderError::NodeNotFound(id)),
+        };
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/flui-rendering/src/pipeline/owner.rs:1406]: Acknowledged — same finding as Codex comment_id=3294224488. Documented as KNOWN MIRI BLOCKER on `layout_dirty_root` rustdoc with the architectural fix path (pre-acquire all subtree `&mut RenderNode` via new `RenderTree::get_subtree_mut` primitive) and gated `#[doc(hidden)]` until the fix lands. Filed as U20.1 follow-up blocking U23s `run_layout` wiring. Not fixing in this PR per scope decision.
